src/Swarm/Game/CESK.hs

```diff
@@ -350,39 +352,22 @@ prettyFrame (FDef x) = "def " ++ from x ++ " = _"
 prettyFrame FExec = "exec _"
 prettyFrame (FBind Nothing t _) = "_ ; " ++ prettyString t
 prettyFrame (FBind (Just x) t _) = from x ++ " <- _ ; " ++ prettyString t
-prettyFrame FImmediate {} = "(_ : cmd a)"
+prettyFrame (FImmediate c w r) = unwords ["immediate@" <> T.unpack (Syntax.syntax $ Syntax.constInfo c), show w, show r]
 prettyFrame (FUpdate loc) = "store@" ++ show loc ++ "(_)"
 prettyFrame FFinishAtomic = "finishAtomic"
 
 --------------------------------------------------------------
 -- Wrappers for functions in FImmediate
---
--- NOTE: we can not use GameState and Robot directly, as it
--- would create a cyclic dependency. The alternative is
--- making CESK, Cont and Frame polymorphic which just muddies
--- the picture too much for one little game feature.
---
--- BEWARE: the types do not follow normal laws for Show and Eq
 --------------------------------------------------------------
 
-newtype WorldUpdate = WorldUpdate
-  { worldUpdate :: World Int Entity -> Either Exn (World Int Entity)
-  }
-
-newtype RobotUpdate = RobotUpdate
-  { robotUpdateInventory :: Inventory -> Inventory
+data WorldUpdate = ReplaceEntity
+  { updatedLoc :: V2 Int64
+  , originalEntity :: Entity
+  , newEntity :: Maybe Entity
   }
+  deriving (Eq, Ord, Show, Generic, FromJSON, ToJSON)
```

Comment:
The default instances do not follow our style, so I would like to customize it a bit and test it before merging.

Reply:
Scratch that, doing it here would blow up the scope and I want to clean up first.